### PR TITLE
feat(CLI): Extract SLPK

### PR DESCRIPTION
**Pull request by dariaterekhovaae**

### CODE REVIEW

File: modules/i3s/src/lib/parsers/parse-zip/buffer-file-provider.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
+import {FileProvider} from './file-provider';
+
+/**
+ * Provides file data using DataView
+ */
+export class BufferFileProvider implements FileProvider {
+  /**
+   * The DataView from which data is provided
+   */
+  file: DataView;
```
Comment (belom88) [inline modules/i3s/src/lib/parsers/parse-zip/buffer-file-provider.ts:10]: Probably, make it private

File: modules/i3s/src/lib/parsers/parse-zip/buffer-file-provider.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  /**
+   * Gets an unsigned 8-bit integer (unsigned byte) at the specified byte offset from the start of the file.
+   * @param offset The offset, in byte, from the start of the file where to read the data.
+   */
+  getUint8(offset: number): Promise<number> {
+    return Promise.resolve(this.file.getUint8(offset));
+  }
+
+  /**
+   * Gets an unsigned 16-bit integer (unsigned byte) at the specified byte offset from the start of the file.
```
Comment (belom88) [inline modules/i3s/src/lib/parsers/parse-zip/buffer-file-provider.ts:25]:
```suggestion
   * Gets an unsigned 16-bit integer at the specified byte offset from the start of the file.
```
16-bit is not byte. Well known name is "short"

File: modules/i3s/src/lib/parsers/parse-zip/buffer-file-provider.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+   * @param offset The offset, in byte, from the start of the file where to read the data.
+   */
+  getUint16(offset: number): Promise<number> {
+    return Promise.resolve(this.file.getUint16(offset, true));
+  }
+
+  /**
+   * Gets an unsigned 32-bit integer (unsigned byte) at the specified byte offset from the start of the file.
```
Comment (belom88) [inline modules/i3s/src/lib/parsers/parse-zip/buffer-file-provider.ts:33]:
```suggestion
   * Gets an unsigned 32-bit integer at the specified byte offset from the start of the file.
```
32-bit are 4 bytes

File: modules/i3s/src/lib/parsers/parse-zip/buffer-file-provider.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+   * @param offset The offset, in byte, from the start of the file where to read the data.
```
Comment (belom88) [inline modules/i3s/src/lib/parsers/parse-zip/buffer-file-provider.ts:34]:
```suggestion
   * @param offset The offset, in bytes, from the start of the file where to read the data.
```

File: modules/i3s/src/lib/parsers/parse-zip/buffer-file-provider.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+   * Gets an unsigned 8-bit integer (unsigned byte) at the specified byte offset from the start of the file.
+   * @param offset The offset, in byte, from the start of the file where to read the data.
+   */
+  getUint8(offset: number): Promise<number> {
+    return Promise.resolve(this.file.getUint8(offset));
+  }
+
+  /**
+   * Gets an unsigned 16-bit integer (unsigned byte) at the specified byte offset from the start of the file.
+   * @param offset The offset, in byte, from the start of the file where to read the data.
```
Comment (belom88) [inline modules/i3s/src/lib/parsers/parse-zip/buffer-file-provider.ts:26]:
```suggestion
   * @param offset The offset, in bytes, from the start of the file where to read the data.
```

File: modules/i3s/src/lib/parsers/parse-zip/buffer-file-provider.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+   */
+  file: DataView;
+
+  constructor(file: DataView) {
+    this.file = file;
+  }
+
+  /**
+   * Gets an unsigned 8-bit integer (unsigned byte) at the specified byte offset from the start of the file.
+   * @param offset The offset, in byte, from the start of the file where to read the data.
```
Comment (belom88) [inline modules/i3s/src/lib/parsers/parse-zip/buffer-file-provider.ts:18]:
```suggestion
   * @param offset The offset, in bytes, from the start of the file where to read the data.
```

File: modules/i3s/src/lib/parsers/parse-zip/buffer-file-provider.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+   * Gets an unsigned 32-bit integer (unsigned byte) at the specified byte offset from the start of the file.
+   * @param offset The offset, in byte, from the start of the file where to read the data.
+   */
+  getUint32(offset: number): Promise<number> {
+    return Promise.resolve(this.file.getUint32(offset, true));
+  }
+
+  /**
+   * returns an ArrayBuffer whose contents are a copy of this file bytes from startOffset, inclusive, up to endOffset, exclusive.
+   * @param startOffsset The offset, in byte, from the start of the file where to start reading the data.
```
Comment (belom88) [inline modules/i3s/src/lib/parsers/parse-zip/buffer-file-provider.ts:42]:
```suggestion
   * @param startOffsset The offset, in bytes, from the start of the file where to start reading the data.
```

File: modules/i3s/src/lib/parsers/parse-zip/buffer-file-provider.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+   * @param endOffset The offset, in byte, from the start of the file where to end reading the data.
```
Comment (belom88) [inline modules/i3s/src/lib/parsers/parse-zip/buffer-file-provider.ts:43]:
```suggestion
   * @param endOffset The offset, in bytes, from the start of the file where to end reading the data.
```

File: modules/i3s/src/lib/parsers/parse-zip/file-provider.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+  /**
+   * Gets an unsigned 32-bit integer (unsigned byte) at the specified byte offset from the start of the file.
+   * @param offset The offset, in byte, from the file of the view where to read the data.
+   */
+  getUint32(offset: number): Promise<number>;
+
+  /**
+   * returns an ArrayBuffer whose contents are a copy of this file bytes from startOffset, inclusive, up to endOffset, exclusive.
+   * @param startOffsset The offset, in byte, from the start of the file where to start reading the data.
+   * @param endOffset The offset, in byte, from the start of the file where to end reading the data.
```
Comment (belom88) [inline modules/i3s/src/lib/parsers/parse-zip/file-provider.ts:26]: Ditto for comments misprints

File: modules/i3s/src/lib/parsers/parse-zip/local-file-header.ts [hunk 10]
```diff
@@ -7,6 +9,10 @@ export type ZipLocalFileHeader = {
    * File name length
    */
   fileNameLength: number;
+  /**
+   * File name
+   */
```
Comment (belom88) [inline modules/i3s/src/lib/parsers/parse-zip/local-file-header.ts:14]:
```suggestion
  /** File name */
```
More compact form

File: modules/tile-converter/src/slpk-converter/slpk-converter.ts [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+type File = {
+  name: string | null;
+  data: ArrayBuffer;
+};
+
+/**
+ * Converter from slpk to i3s
+ */
+export default class SLPKConverter {
+  options: any;
```
Comment (belom88) [inline modules/tile-converter/src/slpk-converter/slpk-converter.ts:33]: Add better type for options

**Review by belom88 — CHANGES_REQUESTED** [review]
* The PR title must be changed;
* I don't agree with naming: "slpk converter". SLPK is part of I3S so "Convert SLPK to I3S" doesn't have sense. I'd prefer "Extract SLPK" and following changes for file names, function names etc.

File: modules/tile-converter/src/slpk-converter/slpk-converter.ts [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+    }
+    return `${parts?.at(1)}/index${parts?.at(2)}`; //TODO add index insert
+  }
+
+  private async unGzip(file: File): Promise<File> {
+    if (/\.gz$/.test(file.name ?? '')) {
+      const decompressedData = await processOnWorker(CompressionWorker, file.data, {
+        compression: 'gzip',
+        operation: 'decompress',
+        _workerType: 'test',
```
[the remark this answers is not in the document]
Comment (dariaterekhovaae) [reply modules/tile-converter/src/slpk-converter/slpk-converter.ts:106]: What worker type should I use? Or where can I find any info about them?

File: modules/tile-converter/src/slpk-converter-cli.ts [hunk 13]
```diff
@@ -4,14 +4,14 @@ import SLPKConverter from './slpk-converter/slpk-converter';
 // import {join} from 'path';
```
Comment (belom88) [inline modules/tile-converter/src/slpk-converter-cli.ts:4]: remove "converter" from naming (file names and variables)

File: modules/tile-converter/src/slpk-converter/slpk-converter.ts [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+  }
+
+  private async writeFile(options: File, outputPath: string): Promise<void> {
+    if (!options.name) {
+      return;
+    }
+    const finalPath = path.join(outputPath, options.name);
+    const dirName = path.dirname(finalPath);
+    const fileName = path.filename(finalPath);
+    await writeFile(dirName, options.data, fileName);
```
Comment (belom88) [inline modules/tile-converter/src/slpk-converter/slpk-converter.ts:101]: Can we do it without `path` usage:
```suggestion
    await writeFile(outputPath, options.data, options.name);
```

**Review by belom88 — CHANGES_REQUESTED** [review]
Hidden comments from previous review are not fixed

File: modules/i3s/src/lib/parsers/parse-zip/local-file-header.ts [hunk 15]
```diff
@@ -21,34 +27,47 @@ export type ZipLocalFileHeader = {
...
-): ZipLocalFileHeader => {
-  const offsets = {
-    COMPRESSED_SIZE_OFFSET: 18,
-    FILE_NAME_LENGTH_OFFSET: 26,
-    EXTRA_FIELD_LENGTH_OFFSET: 28,
-    FILE_NAME_OFFSET: 30
-  };
+  buffer: FileProvider
+): Promise<ZipLocalFileHeader | undefined> => {
+  if (Buffer.from(await buffer.slice(headerOffset, headerOffset + 4)).compare(signature) !== 0) {
```
[the remark this answers is not in the document]
Comment (dariaterekhovaae) [reply modules/i3s/src/lib/parsers/parse-zip/local-file-header.ts:49]: This behavior is tested during the slpk loader browser test

File: modules/tile-converter/test/slpk-converter/slpk-converter.spec.js [hunk 16]
```diff
@@ -0,0 +1,23 @@
+import test from 'tape-promise/tape';
+import SLPKConverter from '../../src/slpk-extractor/slpk-extractor';
+import {isBrowser} from '@loaders.gl/core';
+import {BROWSER_ERROR_MESSAGE} from '../../src/constants';
+import {cleanUpPath} from '../utils/file-utils';
+
+const SLPKUrl = 'modules/i3s/test/data/DA12_subset.slpk';
+const outputUrl = 'data';
+
+test('tile-converter - searchFromTheEnd', async (t) => {
```
Comment (belom88) [inline modules/tile-converter/test/slpk-converter/slpk-converter.spec.js:10]: Could you rename this test as I done unified naming: https://github.com/visgl/loaders.gl/pull/2533/files.
Start from "tile-converter(slpk-extractor)#......"

**Review by belom88 — CHANGES_REQUESTED** [review]

**Review by belom88 — APPROVED** [review]
Some extra comments nice to have

File: modules/i3s/src/lib/parsers/parse-zip/buffer-file-provider.ts [hunk 17]
```diff
@@ -0,0 +1,55 @@
+import {FileProvider} from './file-provider';
+
+/**
+ * Provides file data using DataView
+ */
+export class BufferFileProvider implements FileProvider {
```
Comment (ibgreen) [inline modules/i3s/src/lib/parsers/parse-zip/buffer-file-provider.ts:6]: If this provides a file using DataView, why is it not called `DataViewFileProvider`? `BufferFileProvider` seems to suggest the Node.js `Buffer` class.

File: modules/i3s/src/lib/parsers/parse-zip/buffer-file-provider.ts [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+  getUint32(offset: number): Promise<number> {
+    return Promise.resolve(this.file.getUint32(offset, true));
+  }
+
+  /**
+   * returns an ArrayBuffer whose contents are a copy of this file bytes from startOffset, inclusive, up to endOffset, exclusive.
+   * @param startOffsset The offset, in bytes, from the start of the file where to start reading the data.
+   * @param endOffset The offset, in bytes, from the start of the file where to end reading the data.
+   */
+  slice(startOffsset: number, endOffset: number): Promise<ArrayBuffer> {
```
Comment (ibgreen) [inline modules/i3s/src/lib/parsers/parse-zip/buffer-file-provider.ts:45]:
```suggestion
  slice(startOffset: number, endOffset: number): Promise<ArrayBuffer> {
```

File: modules/i3s/src/lib/parsers/parse-zip/buffer-file-provider.ts [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+   * Gets an unsigned 32-bit integer at the specified byte offset from the start of the file.
+   * @param offset The offset, in bytes, from the start of the file where to read the data.
+   */
+  getUint32(offset: number): Promise<number> {
+    return Promise.resolve(this.file.getUint32(offset, true));
+  }
+
+  /**
+   * returns an ArrayBuffer whose contents are a copy of this file bytes from startOffset, inclusive, up to endOffset, exclusive.
+   * @param startOffsset The offset, in bytes, from the start of the file where to start reading the data.
```
Comment (ibgreen) [inline modules/i3s/src/lib/parsers/parse-zip/buffer-file-provider.ts:42]:
```suggestion
   * @param startOffset The offset, in bytes, from the start of the file where to start reading the data.
```

**Review by ibgreen — APPROVED** [review]
The FileProvider overlaps some other similar code in loader-utils. 

We don't need to align now, but we may want to come back to it.

When working in the browser you'd benefit from BlobProvider. Blobs allow you to read partial files so that you can work with very large files, perfect for SLPKs.

**Pull request merged**